### PR TITLE
Storage: Add mode-specific dual writers

### PR DESCRIPTION
**Pull request by suntala**
**What is this feature?**

Adds dual writers for each of the modes defined in https://docs.google.com/document/d/1_FZ75QYoBqAW_7pC64fPN5GPUbHyRIWmemcO9SDhBJM/edit#heading=h.5sybau7waq2q. Added the Create method as an example. We'll add the other methods progressively in separate PRs. Tests are in progress in a separate PR and will be added progressively as well: https://github.com/grafana/grafana/pull/85190. 

There's some overlap with https://github.com/grafana/grafana/pull/84605 but that PR will remain in draft stage as a PoC.

Additional todos which are not in scope for this issue are listed here: https://github.com/grafana/search-and-storage-team/issues/5. This includes handling failure modes.

**Why do we need this feature?**

Currently when a resource is created or updated via the new API server, k8s-object fields such as labels are not persisted.

**Who is this feature for?**

U.S. developers and users.

**Which issue(s) does this PR fix?**:

Partial https://github.com/grafana/search-and-storage-team/issues/17

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/apiserver/rest/dualwriter_mode2.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	c, err := enrichObject(obj, created)
+	if err != nil {
+		return created, err
+	}
+
+	accessor, err := meta.Accessor(c)
+	if err != nil {
+		return created, err
+	}
+	accessor.SetResourceVersion("")
```
Comment (leonorfmartins) [inline pkg/apiserver/rest/dualwriter_mode2.go:45]: any reason why we are setting the resource version to empty?

File: pkg/apiserver/rest/dualwriter.go [hunk 2]
```diff
@@ -65,8 +65,21 @@ type DualWriter struct {
...
+type DualWriterMode int
+
+const (
+	Mode1 DualWriterMode = iota
+	Mode2
+	Mode3
+	Mode4
+)
+
+var CurrentMode = Mode2
```
Comment (leonorfmartins) [inline pkg/apiserver/rest/dualwriter.go:77]: I would default the currentMode to 2 only when we have all the methods merged, wdyt?

File: pkg/apiserver/rest/dualwriter.go [hunk 3]
```diff
@@ -65,8 +65,21 @@ type DualWriter struct {
...
+
+// #TODO make CurrentMode customisable and specific to each entity
+
+// newDualWriter returns a new DualWriter.
```
Comment (leonorfmartins) [inline pkg/apiserver/rest/dualwriter.go:81]:
```suggestion
// newDualWriter returns a new dualWriter.
```

File: pkg/apiserver/rest/dualwriter.go [hunk 4]
```diff
@@ -8,7 +8,7 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apiserver/pkg/registry/rest"
-	"k8s.io/klog/v2"
```
Comment (leonorfmartins) [inline pkg/apiserver/rest/dualwriter.go:11]: this seems to be the log that's mostly used in grafana: https://github.com/grafana/grafana/tree/main/pkg/infra/log. What do you think of using that instead?

File: pkg/apiserver/rest/dualwriter.go [hunk 5]
```diff
@@ -65,8 +65,21 @@ type DualWriter struct {
...
+	Mode4
+)
+
+var CurrentMode = Mode2
+
+// #TODO make CurrentMode customisable and specific to each entity
+
+// newDualWriter returns a new DualWriter.
+func newDualWriter(legacy LegacyStorage, storage Storage) *DualWriter {
 	return &DualWriter{
```
Comment (leonorfmartins) [inline pkg/apiserver/rest/dualwriter.go:83]:
```suggestion
	return &dualWriter{
	storage: storage,
```

Comment (leonorfmartins) [reply pkg/apiserver/rest/dualwriter.go:83]: it seems weird for a private function to return a public type. I'd rather keep everything public if we need to export it or to keep everything private. Can you think if a scenario of why this should be public? Otherwise let's keep it private

File: pkg/apiserver/rest/dualwriter_mode2.go (hunk 1 above)
Comment (suntala) [reply pkg/apiserver/rest/dualwriter_mode2.go:45]: Addressed this and other comments in the PoC PR: https://github.com/grafana/grafana/pull/84605

File: pkg/apiserver/rest/dualwriter.go (hunk 4 above)
Comment (suntala) [reply pkg/apiserver/rest/dualwriter.go:11]: Addressed similar comment in the other PR: https://github.com/grafana/grafana/pull/84605

File: pkg/apiserver/rest/dualwriter.go (hunk 2 above)
Comment (suntala) [reply pkg/apiserver/rest/dualwriter.go:77]: The current mode 2 behaviour falls back to the generic dual writer behaviour for everything except Create because most methods aren't defined in mode 2. Therefore, it's closer to the generic dual writer right now than when we will merge the remaining methods for mode 2.

One main difference in the Create method for mode 2 vs generic dual writing is that we now expect legacy storage to implement the `rest.Creater` instead of defaulting to just writing to Storage. I think that's a fair requirement.

The other one is that we are now setting additional fields in Storage (annotations and labels) but that shouldn't interfere with LegacyStorage Create.

It's also a safe mode to default to because it will continue reading and writing to LegacyStorage but additionally write to Storage on a best effort basis. The idea is to write to Storage when possible but not impede the LegacyStorage behavior. 

To summarise, dual writer mode 2 seems like the safest dual writer to default to besides the generic one. 

To be on the safer side, what I could do now is switch back to registering the generic dual writer and once we're further along and have more testing in place, we can switch it over.

File: pkg/apiserver/rest/dualwriter.go (hunk 5 above)
Comment (suntala) [reply pkg/apiserver/rest/dualwriter.go:83]: This makes sense. Let's keep this in mind for when we stop relying on the generic dual writer. I'm going to switch back to registering the generic DW temporarily.

File: pkg/apiserver/rest/dualwriter.go (hunk 2 above)
Comment (leonorfmartins) [reply pkg/apiserver/rest/dualwriter.go:77]: good point, so considering this I guess it's fine to merge things as they are, wdyt?

**Review by leonorfmartins — APPROVED** [review]
